### DISCUSSION TITLE
Dedicated Migration Container

### DISCUSSION THREAD

**Discussion by acelinkio** [zoriya/Kyoo#480] [General]
There has been a bit of discussion about how to handle managing databases and the amount of containers.  Taking a few moments to write out some of my thoughts.

**CONTEXT**
Currently Kyoo currently leverages a dedicated container, ghcr.io/zoriya/kyoo_migrations, for managing the database that kyoo_back application connects to.  That container is built using the Kyoo backend source code in the `back` directory.  Looking through the source code, EntityFrameworkCore is used for both generating schema and storing migrations. Those migrations are bundled/packaged into a single purpose artifact dedicated to managing any database setup.

**PRO**
Having separate containers does have some notable benefits including: reducing application runtime permissions, avoiding supporting long startup times due to database management in the application, and overall makes support easier.

**CON**
Dedicated migration container is overkill in most situations, especially with microservice architecture where the schema tends to be only a few tables/relationships.  Each application should be expected to have its own database.  Modern developers are expected to be the DBAs for their application.  Separating out different users/roles/permissions between migration and application tend to be a lot of additional overhead especially for those getting started.

**REALITY**
EFCore does not have a way for handling concurrent migrations.  https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/applying?tabs=dotnet-core-cli#apply-migrations-at-runtime.  So running inside of the application OR as an init container can potentially lead to conflicts.   Other database versioning tools, such a Liquibase, add database locks to prevent that issues.  A dedicated container is only EFCore supported way.  

**SUGGESTIONS**
* Make it clear that both the database and migration containers are owned by one and only one source.  Would encourage that to be reflected in both the database name and migration container name.
```yaml
application: kyoo_back
database: kyoo_back
migration: kyoo_back_rdbmigration
```
* Do NOT have other applications connect directly to another's backend.  Rely upon application's api or message buses for sharing information.  This is extremely important if you ever plan to break out the backend into multiple pieces.
* For ease of use, use the same credentials for both applying the migration and running the application.  Most postgres deployment methods in Docker or Kubernetes do not make it easy to manage users/roles/permissions/databases/schemas.  Advanced users can separate that out if they wish...
* Kubernetes: Use a job for executing the migration container and be comfortable with the application crash looping until reaching eventual consistency.  (Helm pre-upgrade/pre-install hooks are possible, but timing is funky especially with subcharts for postgres/milisearch/rabbitmq.)
* No monolithic migration containers.  Each backend management container should only connect to one and only one resource.  PostgresManagementContainer should only connect to postgres.  CassandraManagementContainer should only connect to cassandra. etc.
* Any back fill operations need to be in their own container separate from migrations.  Back fill workloads are run once and forgotten about.

Taking the jump into a distributed systems is challenging.  So far everything I see in Kyoo is going in the right direction!  Use as many containers as needed for avoiding monoliths.

**Comment by zoriya** [comment]
Thanks for the detailed document, do you think the migration container is the way to go? I made that since I had issues with enums when the back was doing the migrations and tried the recommended way of doing it.

My opinion on the migration container is:

Pro:
- nice on DX when using dev containers with hot reload, since new migrations are not applied while you write them (can be solved by simply stopping the API container while you work on a migration)

Cons:
- Since EFCore builds the whole projects to get migrations, we need to `COPY . .` in the Dockerfile so it needs to build again at every file change (which can be quite slow).
- Can get pretty boilerplate-y to have a migration container per db/app, I have not looked at how meilisearch's indexes migrations should be configured (I believe it's okay to reconfigure it even concurrently so not that important but still something to note). I also want to note that, for now, the transcoder store some metadata (list of keyframes, mediainfo of videos) as json files on a container. I might want to move them to sql at one point, if I do, I think it would be on a new postgres schema.
- EF's migration bundle cannot be customized/edited so adding locks to prevent concurrent runs is not possible.

I also dislike the way EF does migrations with lots of generated c# files and all, but I think we're stuck with it since that's the way the ORM works.

Overall, I think moving back migrations to the api's container might be good, what do you think?

**Reply by acelinkio** [reply]
Moving the migration into the kyoo_backend container is a good idea. That database is pretty lightweight and all transactions should complete in seconds.  Concurrent applies, rollbacks, or compute hungry query optimizations via index/partitioning should not be that big concern.

> I also dislike the way EF does migrations with lots of generated c# files and all, but I think we're stuck with it since that's the way the ORM works.

This is a bit pedantic but EFCore is acting both as an ORM and DatabaseVersioning.  ORMs focus on creating the underlying structures.  Those outputs are then consumed by DatabaseVersioning  toolsto manage all of the incremental steps to get there in a declarative way.  **Honestly if you are confident with the data structures, you could just rely on the ORM to get the database setup in an imperative way to avoid checking in those step by step instructions.**

EFCore is fine as an ORM, but the versioning feels more like an afterthought bolted on.  If you want to continue versioning, that can be moved to another tool like liquibase which only needs sql transactions output from an ORM.

**Reply by zoriya** [reply]
The more I think about this, the more I want to ditch EF's migrations tools. I always thought it was a bad idea to hide sql statements in a bunch of auto-generated c# files.
Having a separate tool would make this easier to add schemas for others services as well, since we could use the same tool/logic to build the other schemas.

I'm going to look at liquibase (since it's the one you mentioned) in the following days/weeks.

**Comment by acelinkio** [comment]
Closing discussion.  v5 choose to keep the migrations inside of the applications and keep it simple.
